### ISSUE TITLE
User Experience Expectation Regarding `XLATensor` Design

### ISSUE THREAD

**Issue by miladm** [pytorch/pytorch#65130]
To enable Dynamic Shape support, is it fair to expect the user to switch their `torch.Tensor` objects to `XLATensor`? If yes, when/how should they transition back and forth between torch.Tensor objects and XLATensor objects?

This `[document](https://docs.google.com/document/d/1u5kJ18HKnoJ-i8shymt__wRrnw-eYII3c-AHUlnky-s/edit?resourcekey=0-THFZXxHHehVBA-oBsLU0Jw#heading=h.9iwsbbqufptx)` summarizes the details of our `XLATensor` concept.

```
class XLATensor(torch.Tensor):
    def expand(self, size):
        if self.dev_ == 'TPU':
            if size.dynamic_size_tensor != None:
                result = self.t_.expand(size.static_size)
                return XLATensor(result.tolist())
```

**Comment by ezyang** [comment]
No, this would not be the expectation. What we'd actually want is if you do something like `torch.randn(sizes, device='xla')`, this would just directly return an XLATensor rather than a Tensor. We would now have an invariant that if a tensor is an XLA tensor, it surfaces in Python as XLATensor.

@albanD is working on a generic mechanism to make this be the case, but in the short term we could probably add some special case logic in the Python binding layer to detect if we have XLA tensors and if so, make them XLATensors instead.

**Comment by ezyang** [comment]
The conditional would go here: https://github.com/pytorch/pytorch/blob/1fec9cd76ba3eaa232e6c2d65d7db6710f6ef629/torch/csrc/autograd/python_variable.cpp#L202-L203
